### ISSUE TITLE
0.14 and 0.15 build error against GCC 13

### ISSUE THREAD

**Issue by devrimgunduz** [dimitri/pgcopydb#672]
Hi,

0.14 and 0.15 fails to build against GCC 13 on Fedora 39. Can you please take a look? No problems on RHEL 9 (GCC11) and others. Thanks!

```
cc -std=c99 -D_GNU_SOURCE -g -I /usr/pgsql-14/include -I /usr/pgsql-14/include/server -I /usr/pgsql-14/include/internal -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Wendif-labels -Wmissing-format-attribute -Wimplicit-fallthrough=3 -Wcast-function-type -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -Wformat -Wall -Werror=implicit-int -Werror=implicit-function-declaration -Werror=return-type -Wno-declaration-after-statement -D_WANT_SEMUN -Wno-missing-braces -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/main/non-common/pgcopydb/main/pgcopydb-0.15/src/bin/pgcopydb/../lib/pg -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/main/non-common/pgcopydb/main/pgcopydb-0.15/src/bin/pgcopydb/../lib/log/src/ -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/main/non-common/pgcopydb/main/pgcopydb-0.15/src/bin/pgcopydb/../lib/subcommands.c/ -I/home/devrim/Documents/Devrim/Projects/reo/pgrpms/rpm/redhat/main/non-common/pgcopydb/main/pgcopydb-0.15/src/bin/pgcopydb/../lib/libs/ -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/main/non-common/pgcopydb/main/pgcopydb-0.15/src/bin/pgcopydb/../lib/parson/ -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/main/non-common/pgcopydb/main/pgcopydb-0.15/src/bin/pgcopydb/../lib/uthash/ -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/main/non-common/pgcopydb/main/pgcopydb-0.15/src/bin/pgcopydb/../lib/jenkins/  blobs.o catalog.o cli_clone_follow.o cli_common.o cli_compare.o cli_copy.o cli_dump.o cli_list.o cli_ping.o cli_restore.o cli_root.o cli_sentinel.o cli_snapshot.o cli_stream.o compare.o copydb.o copydb_schema.o dump_restore.o env_utils.o extensions.o file_utils.o filtering.o follow.o indexes.o ini_implementation.o ld_apply.o ld_replay.o ld_stream.o ld_test_decoding.o ld_transform.o ld_wal2json.o lock_utils.o lsn_tracking.o main.o parsing_utils.o pg_utils.o pgcmd.o pgsql.o pidfile.o progress.o queue_utils.o schema.o sentinel.o sequences.o signals.o snapshot.o string_utils.o summary.o table-data.o vacuum.o lib-log.o lib-commandline.o lib-parson.o lib-snprintf.o lib-strerror.o lib-dumputils.o lib-string_utils.o lib-lookup3.o  -L /usr/pgsql-14/lib -L /usr/pgsql-14/lib -Wl,-z,relro -Wl,--as-needed -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -Wl,--build-id=sha1 -L/usr/lib64 -L/usr/lib64 -Wl,--as-needed -Wl,-rpath,'/usr/pgsql-14/lib',--enable-new-dtags -lpgcommon -lpgport -lselinux -llz4 -lxslt -lxml2 -lpam -lssl -lcrypto -lgssapi_krb5 -lz -lreadline -lm  -lpq -lncurses -lsqlite3 -o pgcopydb
/usr/bin/ld: lib-lookup3.o: relocation R_X86_64_32S against `.rodata' can not be used when making a PIE object; recompile with -fPIE
/usr/bin/ld: failed to set dynamic section sizes: bad value
collect2: error: ld returned 1 exit status
make[3]: *** [Makefile:99: pgcopydb] Error 1
```

**Comment by markwort** [comment]
I see that there is some stuff in the Makefile to run the compiler with `-fpie`, but not `fPIE`.
The link mentioned in the comment (I hoped explaining why these particular flags have been chosen) doesn't lead me anywhere useful, but to a MS login page...

https://github.com/dimitri/pgcopydb/blob/a7889dfa93385c2fcd82de4d9b3c96e48810524a/src/bin/pgcopydb/Makefile#L71-L75

What's the intended use for `USE_SECURITY_FLAGS` ?
Should packagers set that by default when compiling pgcopydb?

In any case I understand that `-fpie` [should be changed](https://gcc.gnu.org/onlinedocs/gnat_ugn/Position-Independent-Executable-PIE-Enabled-by-Default-on-Linux.html) to `-fPIE`, but my understanding in what exactly that changes is a bit limited.
But the hint is quite explicit about using `-fPIE`, so I guess `-fpie` doesn't change the issue.

[Here](https://gcc.gnu.org/onlinedocs/gnat_ugn/Position-Independent-Executable-PIE-Enabled-by-Default-on-Linux.html)'s some reference on `-fpie` and `-fPIE`.

**Comment by dimitri** [comment]
Hey @hanefi I believe you have some experience with both RPM builds and also the Microsoft security/hardening build options. Could you please have a look at that?

**Comment by hanefi** [comment]
It turned out that we had a typo when setting the CFLAGS for some reason. We did not pass the same set of CFLAGS when compiling `lib-lookup3.o` and ended up with a file that is not position independent object code. Hence the errors when linking this object to the rest of the project.

While I was at it, I removed the currently unused environment variables for Microsoft Compliance, and use the required flags for MS standards at all times.
